extend find_fallback_nic() to support FreeBSD

igalic opened the pull request:
this uses `route -n show default` to find the default interface

igalic commented:
this is an extension to https://github.com/canonical/cloud-init/pull/61
to support network data sources

OddBloke commented on cloudinit/net/__init__.py:
```diff
@@ -327,8 +327,49 @@ def is_disabled_cfg(cfg):
+def find_fallback_nic_on(blacklist_drivers=None):
     """Return the name of the 'fallback' network device."""
+    if util.is_FreeBSD():
+        return find_fallback_nic_on_freebsd()
```
Should these differences be captured on the distro class?

igalic replied:
i followed @goneri's example here, but, basically, probably, yes.

goneri replied:
The root of the problem is that most if the functions of `cloudinit/net/__init__.py` assume they are on Linux. So, in an ideal world, we should move all the OS related logic in some kind of `cloud/os/linux.py` and `cloud/os/freebsd.py`.

igalic replied:
from reading https://github.com/canonical/cloud-init/blob/master/cloudinit/distros/__init__.py it would appear this is the Linux os class

goneri commented:
Hi @igalic, thank you for the PR.

My implementation of `get_interfaces_by_mac()` was broken, as a result `distros.freebsd.generate_fallback_config()` was not able to list local interfaces (
https://github.com/goneri/cloud-init/blob/freebsd_net_renderer/cloudinit/distros/freebsd.py#L188-L194)

I just boot a FreeBSD10 without any metadata, and cloud-init has generated the following (correct) entry:

```
ifconfig_vtnet0=DHCP
```

So, I'm sorry for the situation, I close the PR for now. It would be fabulous if you can retry my branch https://github.com/canonical/cloud-init/pull/61.

BTW, I rebased it, and I'm glade I don't get flooded anymore by `get_linux_distro()` error messages :-). Thank you!

goneri commented:
As patiently explain by @igalic, we actually need the function to use `EphemeralIPv4Network`.
```
[16:05] <meena> anyway, i still think we need to implement find_fallback_nic() cuz that's what all the EphmeralNetwork and EphemeralDHCP things use.
[16:10] <goneri> this https://github.com/goneri/cloud-init/blob/freebsd_net_renderer/cloudinit/distros/freebsd.py#L188-L194 is supposed to do the same thing
[16:10] <meena> i'll merge it in and see if that makes it work betterer
[16:11] <goneri> so, from my understanding, if something use find_fallback_nic(), it means it bypass the network renderer, which is a bad idea
[16:12] <meena> yesno…
[16:14] <meena> it'll still use the network renderer for the real configuration.
[16:14] <goneri> indeed, I miss-understood the purpose of EphemeralIPv4Network()
[16:15] <goneri> I just read the documentation
[16:15] <meena> it's so you bring up enough network to make a connection to the metadata service 
```

igalic committed:
extend find_fallback_nic() to support FreeBSD

this uses `route -n show default` to find the default interface

goneri commented:
Thanks for the rebase, could you also fix the unit-tests?

goneri commented on cloudinit/net/__init__.py:
```diff
@@ -329,6 +329,47 @@ def is_disabled_cfg(cfg):
+
+    we're interested in ``interface: vtnet0``.
+    """
+    route, stderr = util.subp(['route', '-n', 'show', 'default'])
```
The network is supposed to be down (unless I miss something). Can we already use `route`?

igalic replied:
indeed.
so this is then basically a special case, if we already have some network connection

hence also https://github.com/canonical/cloud-init/pull/69

We also need to cover the case of no network.

the Linux code branch does this with a long dance to figure out which interfaces are connected, and i'm sure We can emulate at in a bsd appropriate fashion

perhaps we can even build a parser to pull all that data out into a dict and provide all the information necessary, easily accessible

igalic replied:
i think this is reasonable, and we should go for it

```python
ifcfg, stderr = util.subp(['ifconfig', '-a'])
# handle stderr
netcfg = {}
ifcfgiter = ifcfg.split("\n")
for l in ifcfgiter:
    m = l.search(r"^(?P<netif>\S*): flags=\d+<?P<flags>([\w+,]+,\w+)> .*")
    if m is not None:
        netif = m.group("netif")
        flags = m.group("flags")
        netcfg[netif] = { 'flags': flags.lowercase().split(",") }
        l = next(ifcfgiter) # this necessary?
        for l in ifcfgiter:
            # parse body:
            # inet, inet6
            # mac / ether address
            # group(s)
            # members of vlans or bridges
            # status
            # fill netcfg dict
```

goneri replied:
Or just the first line of `ifconfig -l -u ether` :-)

goneri commented on cloudinit/net/__init__.py:
```diff
@@ -329,6 +329,47 @@ def is_disabled_cfg(cfg):
+        maybe_netif = via_re.match(l)
+
+        if maybe_netif is not None:
+            return maybe_netif.group("netif")
```
You can get a list of the interfaces with `get_interfaces_by_mac().keys()`.

igalic replied:
to see if we got a valid one?

goneri replied:
At this point, I would go with a naive: Pick the first one with a link up.

igalic replied:
so, what would be the correct interface in this example output:

https://github.com/canonical/cloud-init/blob/97e2b5de7ea4dc861b10914908f40905d837c330/tests/data/netinfo/freebsd-ifconfig-output

?

goneri replied:
well `vtnet0`.

igalic committed:
simplify find_fallback_nic

route implies that the interfaces are up.
We'll simply use the ethernet first interface that's up.

goneri commented on cloudinit/net/__init__.py:
```diff
@@ -342,30 +342,15 @@ def find_fallback_nic_on_freebsd(blacklist_drivers=None):
-        if maybe_netif is not None:
-            return maybe_netif.group("netif")
-
-    return None
```
Nitpicking, but this looks like superfluous `try/except`. You can get the same result with just:

```
if stdout:
    return stdout.split()[0]
```

igalic committed:
address review: no try/except, just check stdout!

igalic committed:
address pylint

goneri commented:
Thanks @igalic <3.
